central/role/datastore/datastore.go

```diff
@@ -22,8 +22,8 @@ type DataStore interface {
 	GetAllPermissionSets(ctx context.Context) ([]*storage.PermissionSet, error)
 	CountPermissionSets(ctx context.Context) (int, error)
 	AddPermissionSet(ctx context.Context, permissionSet *storage.PermissionSet) error
-	UpdatePermissionSet(ctx context.Context, permissionSet *storage.PermissionSet) error
-	RemovePermissionSet(ctx context.Context, id string) error
+	UpdatePermissionSet(ctx context.Context, permissionSet *storage.PermissionSet, requestOrigin storage.Traits_Origin) error
+	RemovePermissionSet(ctx context.Context, id string, requestOrigin storage.Traits_Origin) error
```

Comment:
I wonder how easy it would be as a client of this datastore now.

For people that do not know the constraints on the `Traits_Origin`, they would currently have to be careful when calling this that they pass in `storage.Traits_IMPERATIVELY`, as otherwise they would be allowed to modify declarative configurations.
Seeing as `storage.Traits_IMPERATIVELY` is the default value of the enum, since by default we assume all resources are imperatively created ones, this makes it even more confusing to me.

We might think about generalizing this "optionally setting the request origin" problem. I do like options for this, but I'm not sure whether they are overkill (i.e. shared options for the declarative config that overwrite the default origin used, which would be `storage.Traits_IMPERATIVELY`). Alternatively, we would also add the request origin as a variadic args (`...`), allowing people to _optionally_ specify the origin (with the caveat that we now can allow removals of both imperative / declarative origins for clients, which might be something beneficial at some point).

Reply:
> shared options for the declarative config that overwrite the default origin used, which would be storage.Traits_IMPERATIVELY

What do you mean by "shared options"?

Here are the alternative solutions that I see. All of them would not require any change for the clients who work with imperative resources:
1. Introduce functions like `RemoveDeclarativePermissionSet` and `UpdateDeclarativePermissionSet`
2. Introduce variadic args
3. Pass request origin within the context